pkg/types/validation/installconfig.go

```diff
@@ -744,3 +746,30 @@ func validateCapabilities(c *types.Capabilities, fldPath *field.Path) field.Erro
 	}
 	return allErrs
 }
+
+// validateFeatureSet returns an error if a gated feature is used without opting into the feature set.
+func validateFeatureSet(c *types.InstallConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	validFeatureSets := func() []string {
+		v := []string{}
+		for n := range configv1.FeatureSets {
+			v = append(v, string(n))
+		}
+		return v
+	}()
+
+	if !sets.NewString(validFeatureSets...).Has(string(c.FeatureSet)) {
+		allErrs = append(allErrs, field.NotSupported(field.NewPath("featureSet"), c.FeatureSet, validFeatureSets))
+	}
+
+	if c.FeatureSet != configv1.TechPreviewNoUpgrade {
+		errMsg := "the TechPreviewNoUpgrade feature set must be enabled to use this field"
```

Comment:
nit: you can defer this error message until you're inside the `NetworkProjectID` `if` block below.

Reply:
ah, never mind.  I see this is a genric error message that could be shared with future tech-preview properties.